packages/core-app-api/src/lib/AuthConnector/DefaultAuthConnector.ts

```diff
@@ -77,12 +81,37 @@ export class DefaultAuthConnector<AuthSession>
       provider,
       joinScopes = defaultJoinScopes,
       oauthRequestApi,
+      usePopup,
       sessionTransform = id => id,
     } = options;
 
     this.authRequester = oauthRequestApi.createAuthRequester({
       provider,
-      onAuthRequest: scopes => this.showPopup(scopes),
+      onAuthRequest: async scopes => {
+        if (!usePopup) {
+          // modal before redirect
+          const scope = this.joinScopesFunc(scopes);
+          const redirectUrl = await this.buildUrl('/start', {
+            scope,
+            origin: window.location.origin,
+            redirectUrl: window.location.href,
+            authType: 'redirect',
+          });
+
+          if (provider.hasOwnProperty('provider_id')) {
+            // set the sign in provider here
+            localStorage.setItem(
+              '@backstage/core:SignInPage:provider',
+              provider.provider_id!,
+            );
+          }
+
+          window.location.href = redirectUrl;
+          // we need to return to exit function or else popup occurs
+          return Promise.resolve({} as AuthSession);
```

Comment:
I think the best thing to return here is a promise that never resolves. That way execution halts until the page has been redirected and you won't get any strange side effects.

```ts
return new Promise(() => {})
```

Reply:
One of the impacts of having the promise that never returns is that it is very tricky to unit test. Executing this method and triggering the redirect will time out a unit test waiting for a response.